File: 94.py
# LeetCode 94 Binary Tree Inorder Traversal
# Reference:

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        
        # Recursive
        res = []

        def inorder(root):
            if not root:
                return
            
            inorder(root.left)
            res.append(root.val)
            inorder(root.right)
        
        inorder(root)
        return res
    
    def inorderTraversal2(self, root: Optional[TreeNode]) -> List[int]:

        # Iterative
        res = []
        stack = []
        cur = root

        while cur or stack:
            while cur:  # cur is not null
                stack.append(cur)
                cur = cur.left

            cur = stack.pop()
            res.append(cur)
            cur = cur.right
        return res

File: 94.java
// LeetCode 94 Binary Tree Inorder Traversal
//Reference: https://leetcode.com/problems/binary-tree-inorder-traversal/solutions/3169549/solution/?envType=featured-list&envId=top-interview-questions?envType=featured-list&envId=top-interview-questions
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    private List<Integer> res = new ArrayList<>();


    public List<Integer> inorderTraversal(TreeNode root) {
        traverse(root);
        return res;
    }

    
    private void traverse(TreeNode root){
        if (root == null){ return ;}

        traverse(root.left);
        res.add(root.val);
        traverse(root.right);
    }
}